core/chains/evm/types/models.go

```diff
@@ -90,10 +94,49 @@ func (h *Head) BlockDifficulty() *big.Int {
 
 // EarliestInChain recurses through parents until it finds the earliest one
 func (h *Head) EarliestInChain() *Head {
-	for h.Parent != nil {
-		h = h.Parent
+	var earliestInChain *Head
+	h.forEach(func(h *Head) bool {
+		earliestInChain = h
+		return true
+	})
+	return earliestInChain
+}
+
+// forEach - go through the chain linked list. Stops when do returns false or h is nil
+func (h *Head) forEach(do func(h *Head) bool) {
+	cur := h
+	fast := h // use Floyd's Cycle-Finding Algo
+	start := time.Now()
+	for cur != nil {
+		if !do(cur) {
+			return
+		}
+
+		if fast != nil {
+			skip := fast.Parent.Load()
+			if skip != nil {
+				fast = skip.Parent.Load()
+			} else {
+				fast = nil
+			}
+		}
+
+		cur = cur.Parent.Load()
+
+		// if there is a circular reference eventually `fast` will catch up with `cur`.
+		// As .Parent can be modified by another goroutine, it's possible that
+		// cycle will be created after `fast` reached end of the list. To prevent infinite loop, we use duration based validation
+		if (cur != nil && cur == fast) || time.Since(start) > 10*time.Second {
+			// we should log the issue instead of panicking to avoid killing whole node.
+			// we can switch to panic, after full transition to loops.
+			lggr, _ := logger.New()
+			logger.Sugared(lggr).Criticalw("Detected cycle in chain structure. Aborting iteration.",
```

Comment:
We cannot construct new loggers arbitrarily. You will either have to inject it, or return errors to be logged from a higher level.

Reply:
Hmm, we do cycle detection when actually setting `Parent`, so these checks are not required.
Returning an error will break many method signatures, and injecting the logger does not seem like a clean solution either.
Dropping the check seems like the best option.